_source/_posts/2022-08-24-state-of-project-loom.md

```diff
@@ -0,0 +1,261 @@
+- Imagine that `updateInventory()` is an expensive long-running operation and `updateOrder()` threw an error. The `handleOrder()` task will be blocked on `inventory.get()` even though `updateOrder()` threw an error. Ideally, we would like the `handleOrder()` task to cancel `updateInventory()` when a failure occurs in `updateOrder()` so that we are not wasting time.
+- If the thread executing `handleOrder()` is interrupted, the interruption is not propagated to the subtasks. In this case `updateInventory()` and `updateOrder()` will leak and continue to run in the background.
+
+For these situations, we would have to carefully write workarounds and failsafe, putting all the burden on the developer.
+
+We can achieve the same functionality with structured concurrency using the code below.
+
+```java
+void handleOrder() throws ExecutionException, InterruptedException {
+    try (var scope = new StructuredTaskScope.ShutdownOnFailure()) {
+        Future<Integer> inventory = scope.fork(() -> updateInventory());
+        Future<Integer> order = scope.fork(() -> updateOrder());
+
+        scope.join();           // Join both forks
+        scope.throwIfFailed();  // ... and propagate errors
+
+        // Here, both forks have succeeded, so compose their results
+        System.out.println("Inventory " + inventory.resultNow() + " updated for order " + order.resultNow());
+    }
+}
+```
+
+Unlike the previous sample using `ExecutorService`, we can now use `StructuredTaskScope` to achieve the same result while confining the lifetimes of the subtasks to the lexical scope, in this case, the body of the `try-with-resources` statement. The code is much more readable, and the intent is also clear. `StructuredTaskScope` also ensures the below behavior automatically
+
+- **Error handling with short-circuiting** — If either the `updateInventory()` or `updateOrder()` fails, the other is cancelled unless its already completed. This is managed by the cancellation policy implemented by `ShutdownOnFailure()`; other policies are possible.
+
+- **Cancellation propagation** — If the thread running `handleOrder()` is interrupted before or during the call to `join()`, both forks are canceled automatically when the thread exits the scope.
+
+- **Observability** — A thread dump would clearly display the task hierarchy, with the threads running `updateInventory()` and `updateOrder()` shown as children of the scope.
+
+## State of Project Loom
+
+The project started in 2017 and has undergone many changes and proposals. Virtual threads were initially called fibers, but later on, it was renamed to avoid confusion. Today with Java 19 getting closer to release, the project has delivered two features discussed above. One as a preview and another as an incubator. Hence the path to stabilization of the features should be more precise.
```

Comment:
```suggestion
The Loom project started in 2017 and has undergone many changes and proposals. Virtual threads were initially called fibers, but later on, they were renamed to avoid confusion. Today with Java 19 getting closer to release, the project has delivered two features discussed above. One as a preview and another as an incubator. Hence the path to stabilization of the features should be more precise.
```